### PR TITLE
11 | create feature get all tasks with pagination

### LINKED ISSUE #11: Get all Tasks with pagination

### PR DESCRIPTION
**Pull request by kyBa-Le**

### CODE REVIEW

File: src/main/java/com/example/todolist/API/Restful/RestController/TaskController.java [hunk 1]
```diff
@@ -58,4 +63,29 @@ public ResponseEntity<?> getTaskById(@PathVariable("id") String id) {
...
+                        task.getDescription(),
+                        task.getUserId()
+                );
+                taskResponses.add(response);
+            });
+
+            Page<TaskResponse> pageTaskResponses = new PageImpl<>(taskResponses, paging, pageTasks.getTotalElements());
+            return ResponseEntity.status(200).body(new ResponseWithData<>("",pageTaskResponses));
+        } catch (IllegalArgumentException e) {
+            return ResponseEntity.status(400).body(e.getMessage());
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/TaskController.java]: return ResponseEntity.status(400).body(new Response(e.getMessage()));

File: src/main/java/com/example/todolist/Infrastructure/Repository/TaskRepository.java [hunk 2]
```diff
@@ -1,13 +1,20 @@
...
+import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.Query;
 import org.springframework.data.repository.PagingAndSortingRepository;
 
 public interface TaskRepository extends PagingAndSortingRepository<Task, String> {
     void save(Task task);
     Task findById(String id);
 
+    @Override
+    @NonNull
```
Comment (vultpham) [inline src/main/java/com/example/todolist/Infrastructure/Repository/TaskRepository.java]: what is the purpose of @NonNull here?

File: src/main/java/com/example/todolist/API/Restful/RestController/TaskController.java [hunk 3]
```diff
@@ -58,4 +63,29 @@ public ResponseEntity<?> getTaskById(@PathVariable("id") String id) {
...
+                        task.getTitle(),
+                        task.getDescription(),
+                        task.getUserId()
+                );
+                taskResponses.add(response);
+            });
+
+            Page<TaskResponse> pageTaskResponses = new PageImpl<>(taskResponses, paging, pageTasks.getTotalElements());
+            return ResponseEntity.status(200).body(new ResponseWithData<>("",pageTaskResponses));
+        } catch (IllegalArgumentException e) {
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/TaskController.java]: move the try-catch to GlobalException

File: src/main/java/com/example/todolist/API/Restful/RestController/TaskController.java [hunk 4]
```diff
@@ -58,4 +63,29 @@ public ResponseEntity<?> getTaskById(@PathVariable("id") String id) {
         );
         return ResponseEntity.status(200).body(new ResponseWithData<>("",taskResponse));
     }
+
+    @GetMapping("")
+    public ResponseEntity<?> getAllTasks(@RequestParam(defaultValue = "0", name = "page") Integer pageNo,
+                                         @RequestParam(defaultValue = "10", name = "size") Integer pageSize) {
+        try {
+            var paging = PageRequest.of(pageNo, pageSize);
+            var pageTasks = taskRepository.findAll(paging);
```
Comment (vultpham) [inline src/main/java/com/example/todolist/API/Restful/RestController/TaskController.java]: tasks instead of pageTasks

**Review by vultpham — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/todolist/Infrastructure/Repository/TaskRepository.java (hunk 2 above)
Comment (kyBa-Le) [reply src/main/java/com/example/todolist/Infrastructure/Repository/TaskRepository.java]: It relies on the warning that: "Not annotated method overrides method annotated with @NonNullApi".

**Review by vultpham — APPROVED** [review]

**Pull request merged**